README.md
GARLIC
======

[![CircleCI](https://circleci.com/gh/reznikmm/garlic.svg?style=svg)](https://circleci.com/gh/reznikmm/garlic)

> GNAT Library for Ada Distributed Environment (garlic)

This is release of the Ada 95 distributed systems annex
for GNAT. New features are described in file NEWS.

Contents.
---------
   - Introduction: Ada 95 & Distributed Systems.
   - Installation, Documentation & Examples.
   - What Does It Run on.
   - Downloading GLADE.
   - Commercial Support.
   - Bug Reports & Technical Questions.
   - Submitting Bug Fixes.
   - GLADE History.
   - Contributions.

------------------------------------------------------------------------------ 

Additional information can be found in files INSTALL and NEWS as well
as in the 'Doc' directory.

Introduction: Ada 95 & Distributed Systems.
-------------------------------------------

Ada 95 is the first general-purpose language to provide a standard
distributed programming paradigm. By combining the distributed and
object-oriented features of Ada 95, it is possible to create an
application where objects are physically distributed over a network of
machines, without having to interface to any low-level communication
layer.  Likewise, by combining the distributed and real-time
capabilities of Ada 95, it is possible to design applications which
meet real-time constraints in a distributed environment.

GLADE is an implementation of the Distributed Systems Annex for the
GNAT compiler. GLADE stands for GNAT Library for Ada Distributed
Execution. It is an complete environment for developping distributed
applications that includes: 

     - a complete PCS (Partition Communication System)
     - a simple partition description language & tool
     - utilities to build and start distributed applications

The caller and receiver stubs needed by the PCS are generated by the
GNAT compiler using special flags. The partitioning tool is
responsible for calling GNAT with the appropriate flags.

Installation, Documentation & Examples.
---------------------------------------

To see how to install the system please read file
[install_guide.md](Doc/install_guide.md)
in Doc directory. Please look under the "Doc" directory for
[documentation](Doc/glade_ug.md) on
how to use GLADE and some more hints on installation possibilities.
Directory "Examples" contains some examples on how to program a
distributed application in Ada 95.

If you have a problem making your partitions connect with each other,
please check that:

  - You can use rsh from any machine involved in the distributed
    application to any other.

  - Your /etc/hosts file (on Unix systems) does not associate your
    machine name with the 127.0.0.1 address; be aware that some systems
    come misconfigured by default.

What does it run on.
--------------------

This release is fully supported on:
  - PC/Linux
  - Windows NT

It may work on other platforms too.

Downloading GLADE.
------------------

We don't provide binary distribution for now.

Bug Reports & Technical Questions.
----------------------------------

Please fill an issue on github.

GLADE History.
--------------

GLADE has been jointly developed by Ada Core Technologies and a research
team from the École Nationale Supérieure des Télécommunications (ENST).
The ENST team is composed of:
- Laurent Pautet (pautet@inf.enst.fr),
- Samuel Tardieu (sam@inf.enst.fr).

Contributions.
---------------

GLADE filtering is part of a free contribution (04/97) from
Prof. A. Strohmeier, T. Wolf and J. Kienzle (EPFL, Lausanne,
Switzerland). Extra filters are available on:

      http://lglwww.epfl.ch/Ada/filters.html

The basic principles behind this contribution for filtering is
described in the following paper: J. Kienzle, Th. Wolf, A. Strohmeier:
"Secure Communication in Distributed Ada", Proceedings of Ada-Europe
'96, Montreux, Switzerland, June 1996, published as Lecture Notes in
Computer Science 1088, pp. 198 - 210, Springer 1996.

GLADE replay features is part of a free contribution (04/97) from
[D. Neri](mailto:d92-dne@nada.kth.se),
KTH (Royal Institute of Technology, Stockholm, Sweden).



Doc/ChangeLog
2020-02-08  Maxim Reznik <reznikmm@gmail.com>

	* glade_ug.md: Convert texi to Markdown
	* *.svg: Convert *.fig to SVG
	
2005-02-24  Laurent Pautet  <pautet@noos.fr>

	* glade_ug.texi, gfdl.texi: fix inconsistencies in texinfo nodes.

2003-12-11  Samuel Tardieu  <sam@rfc1149.net>

	* glade_ug.texi: Change two instances of Wait_Until_Restart into
	Block_Until_Restart.

2003-12-07  Laurent Pautet  <pautet@enst.fr>

	* glade_ug.texi: Update documentation about priority policies and
	remote shell use.

2003-11-12  Samuel Tardieu  <sam@rfc1149.net>

	* glade_ug.texi: Add proper reference to the GFDL.

2003-01-31  Laurent Pautet  <pautet@arlequin.enst.fr>

	* glade_ug.texi: Fix spelling mistakes

2002-02-13  Laurent Pautet  <pautet@arlequin.enst.fr>

	* glade_ug.texi: Include gfdl.texi and fix some spacing
	problems. Fix copyright as well.

2000-05-26  Laurent Pautet  <pautet@inf.enst.fr>

	* user_guide.html: Remove RCS keyword.

2000-05-24  Laurent Pautet  <pautet@inf.enst.fr>

	* glade_ug.texi: Add info on new features (passive partitions,
	multi protocols, multi storages).

	* Makefile: Add glade_ug.texi to release.

	* myconfig.cfg: Update configuration because of syntax changes.

2000-03-01  Laurent Pautet  <pautet@scapin.enst.fr>

	* glade_ug.texi, install_guide.txt: Replace @gnat_version@ and
 	@glade_version@ with __gnat_version__ and __glade_version__.

1999-08-23  Laurent Pautet  <pautet@scapin.enst.fr>

	* glade_ug.texi: Include Ed's suggestions (second wave of comments).

1999-08-20  Laurent Pautet  <pautet@scapin.enst.fr>

	* glade_ug.texi, all-srcs.ada: Include Ed's suggestions.

1999-07-01  Samuel Tardieu  <sam@inf.enst.fr>

	* Makefile: Do not assume that the current directory is in the
	user's PATH.

1999-01-07  Laurent Pautet  <pautet@scapin.enst.fr>

	* glade_ug.texi: Add sections for gnatdist features.

1999-01-06  Laurent Pautet  <pautet@scapin.enst.fr>

	* glade_ug.texi: Add several empty subsections on configuration
	language features.

1999-01-06  Samuel Tardieu  <sam@inf.enst.fr>

	* Makefile: Cleanup to avoid useless operations.

	* texinfo.tex: Add support for pictures in PDF files. Sent to the
	maintainer of texinfo.tex/texi2dvi/texi2pdf for integration.

	* glade_ug.texi: Make some changes.

	* all-srcs.ada: Likewise.

1999-01-06  Laurent Pautet  <pautet@scapin.enst.fr>

	* glade_ug.texi: Add part of gnatdist documentation.

1999-01-05  Laurent Pautet  <pautet@scapin.enst.fr>

	* glade_ug.texi: Include user guide.

1999-01-01  Laurent Pautet  <pautet@scapin.enst.fr>

	* glade_ug.texi: More on tutorial and DSA/CORBA comparison.

	* full-ex.fig, xe-arch.fig, corba-arch.fig : Figures
	for full example, DSA architecture and CORBA architecture.

	* full-ex.fig.txt, xe-arch.fig.txt, corba-arch.fig.txt :
	ASCII versions of the previous figures.
	
 	* adjustbb, corba.ssh, Makefile : Script to adjust wrong
	bounding box, a format for a2ps and Makefile.
        
   	* naming.idl, all-srcs.ada : IDL example and all Ada sources in a
	chopable format.

1998-12-20  Laurent Pautet  <pautet@scapin.enst.fr>

	* user_guide.txt: A section on how to trace what is going on
	inside GLADE. How to set S_RPC and S_PARINT.

1998-08-31  Laurent Pautet  <pautet@scapin.enst.fr>

	* user_guide.txt (Note): Add short documentation on reconnection
 	policy. Update Restrictions section.

1997-12-16  Samuel Tardieu  <sam@inf.enst.fr>

	* glade_ug.texi (Top): Remove stubs-related nodes.

	* user_guide.txt: Remote note on stubs now that they are not
	generated anymore.

1997-04-13  Laurent Pautet <pautet@dorine.enst.fr>
        
        * glade_ug.texi: Starting GLADE User Guide (texinfo). user_guide.html
        should be removed.
        
1997-04-10  Laurent Pautet <pautet@dorine.enst.fr>
        
        * user_guide.txt: Add documentation on the default installed
        filter and the default used filter.
        
1997-02-12  Laurent Pautet <pautet@lglsun11.epfl.ch>
        
        * user_guide.txt: Building a shell script is not the
        default. Directory hierarchy has been changed (everything is under
        directory dsa). Remove obsolete warnings on exceptions.
        
1997-02-11  Laurent Pautet <pautet@inf.enst.fr>
        
        * user_guide.txt, install_guide.txt: Change references to mailing
        lists.
        
1996-11-29  Samuel Tardieu <sam@inf.enst.fr>
        
        * user_guide.txt: Documented the Ada starter.
        
1996-11-27  Laurent Pautet <pautet@inf.enst.fr>
        
        * install_guide.txt: New file.
        
1996-11-23  Laurent Pautet <pautet@quasimodo.enst.fr>
        
        * user_guide.txt: Description of :
        Partition'<Attribute>    ... (apply to all partitions)
        procedure <Main>         ... (declare a main subprorgam; see below)
        <Partition>'Main         ... (use specific partition main subprogr.)
        <Partition>'Command_Line ... (add arg. on command line)
        pragma Version           ... (don't check consistency version)
        pragma Boot_Server       ... (use specific invocation key)
        
        Mention possible execution in heterogeneous sys.
        Remove shell restrictions when Ada starter was requested.
        
1996-11-07  Samuel Tardieu <sam@inf.enst.fr>
        
        * user_guide.txt: Description of --nolaunch and addition of
        --slave.
        
1996-07-05  Samuel Tardieu <sam@inf.enst.fr>
        
        * user_guide.txt: Franco's changes.
        
1996-05-08  Samuel Tardieu <sam@inf.enst.fr>
        
        * gnatdist.txt: Added Garlic and Gnatdist version.
        
1996-05-05  Samuel Tardieu <sam@inf.enst.fr>
        
        * gnatdist.txt: Boot_Server -> Boot_Server.
        
1996-04-22  Samuel Tardieu <sam@inf.enst.fr>
        
        * gnatdist.txt: Updated.
        


pmwiki/scripts/feeds.php
<?php if (!defined('PmWiki')) exit();
/*  Copyright 2005-2007 Patrick R. Michaud (pmichaud@pobox.com)
    This file is part of PmWiki; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published
    by the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.  See pmwiki.php for full details.

    This script provides a number of syndication feed and xml-based 
    metadata options to PmWiki, including Atom, RSS 2.0, RSS 1.0 (RDF), 
    and the Dublin Core Metadata extensions.  This module is typically
    activated from a local configuration file via a line such as

      if ($action == 'atom') include_once("$FarmD/scripts/feeds.php");
      if ($action == 'dc') include_once("$FarmD/scripts/feeds.php");

    When enabled, ?action=atom, ?action=rss, and ?action=rdf produce
    syndication feeds based on any wikitrail contained in the page,
    or, for Category pages, on the pages in the category.  The feeds
    are generated using pagelist, thus one can include parameters such
    as count=, list=, order=, etc. in the url to adjust the feed output.

    ?action=dc will normally generate Dublin Core Metadata for the 
    current page only, but placing a group=, trail=, or link= argument 
    in the url causes it to generate metadata for all pages in the
    associated group, trail, or backlink.

    There are a large number of customizations available, most of which
    are controlled by the $FeedFmt array.  Elements $FeedFmt look like

        $FeedFmt['atom']['feed']['rights'] = 'All Rights Reserved';

    where the first index corresponds to the action (?action=atom),
    the second index indicates a per-feed or per-item element, and
    the third index is the name of the element being generated.
    The above setting would therefore generate a
    "<rights>All Rights Reserved</rights>" in the feed for
    ?action=atom.  If the value of an entry begins with a '<',
    then feeds.php doesn't automatically add the tag around it.
    Elements can also be callable functions which are called to
    generate the appropriate output.

    For example, to set the RSS 2.0 <author> element to the
    value of the last author to modify a page, one can set 
    (in local/config.php):

        $FeedFmt['rss']['item']['author'] = '$LastModifiedBy';

    To use the RSS 2.0 <description> element to contain the
    change summary of the most recent edit, set

        $FeedFmt['rss']['item']['description'] = '$LastModifiedSummary';

    Feeds.php can also be combined with attachments to support
    podcasting via ?action=rss.  Any page such as "PageName"
    that has an mp3 attachment with the same name as the page
    ("PageName.mp3") will have an appropriate <enclosure> element
    in the feed output.  The set of allowed attachments can be
    extended using the $RSSEnclosureFmt array:

        $RSSEnclosureFmt = array('{$Name}.mp3', '{$Name}.mp4');

    References:
      http://www.atomenabled.org/developers/syndication/
      http://dublincore.org/documents/dcmes-xml/
      http://en.wikipedia.org/wiki/Podcasting
*/

## Settings for ?action=atom
SDVA($FeedFmt['atom']['feed'], array(
  '_header' => 'Content-type: text/xml; charset="$Charset"',
  '_start' => '<?xml version="1.0" encoding="$Charset"?'.'>
<feed xmlns="http://www.w3.org/2005/Atom">'."\n",
  '_end' => "</feed>\n",
  'title' => '$WikiTitle',
  'link' => '<link rel="self" href="{$PageUrl}?action=atom" />',
  'id' => '{$PageUrl}?action=atom',
  'updated' => '$FeedISOTime',
  'author' => "<author><name>$WikiTitle</name></author>\n",
  'generator' => '$Version',
  'logo' => '$PageLogoUrl'));
SDVA($FeedFmt['atom']['item'], array(
  '_start' => "<entry>\n",
  'id' => '{$PageUrl}',
  'title' => '{$Title}',
  'updated' => '$ItemISOTime',
  'link' => "<link rel=\"alternate\" href=\"{\$PageUrl}\" />\n",
  'author' => "<author><name>{\$LastModifiedBy}</name></author>\n",
  'summary' => '{$Description}',
  'category' => "<category term=\"\$Category\" />\n",
  '_end' => "</entry>\n"));

## Settings for ?action=dc
SDVA($FeedFmt['dc']['feed'], array(
  '_header' => 'Content-type: text/xml; charset="$Charset"',
  '_start' => '<?xml version="1.0" encoding="$Charset"?'.'>
<!DOCTYPE rdf:RDF PUBLIC "-//DUBLIN CORE//DCMES DTD 2002/07/31//EN"
    "http://dublincore.org/documents/2002/07/31/dcmes-xml/dcmes-xml-dtd.dtd">
<rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
         xmlns:dc="http://purl.org/dc/elements/1.1/">'."\n",
  '_end' => "</rdf:RDF>\n"));
SDVA($FeedFmt['dc']['item'], array(
  '_start' => "<rdf:Description rdf:about=\"{\$PageUrl}\">\n",
  'dc:title' => '{$Title}',
  'dc:identifier' => '{$PageUrl}',
  'dc:date' => '$ItemISOTime',
  'dc:type' => 'Text',
  'dc:format' => 'text/html',
  'dc:description' => '{$Description}',
  'dc:subject' => "<dc:subject>\$Category</dc:subject>\n",
  'dc:publisher' => '$WikiTitle',
  'dc:author' => '{$LastModifiedBy}',
  '_end' => "</rdf:Description>\n"));

## RSS 2.0 settings for ?action=rss
SDVA($FeedFmt['rss']['feed'], array(
  '_header' => 'Content-type: text/xml; charset="$Charset"',
  '_start' => '<?xml version="1.0" encoding="$Charset"?'.'>
<rss version="2.0" xmlns:dc="http://purl.org/dc/elements/1.1/">
<channel>'."\n",
  '_end' => "</channel>\n</rss>\n",
  'title' => '$WikiTitle | {$Group} / {$Title}',
  'link' => '{$PageUrl}?action=rss',
  'description' => '{$Group}.{$Title}',
  'lastBuildDate' => '$FeedRSSTime'));
SDVA($FeedFmt['rss']['item'], array(
  '_start' => "<item>\n",
  '_end' => "</item>\n",
  'title' => '{$Group} / {$Title}',
  'link' => '{$PageUrl}',
  'description' => '{$Description}',
  'dc:contributor' => '{$LastModifiedBy}',
  'dc:date' => '$ItemISOTime',
  'pubDate' => '$ItemRSSTime',
  'enclosure' => 'RSSEnclosure'));

## RDF 1.0, for ?action=rdf
SDVA($FeedFmt['rdf']['feed'], array(
  '_header' => 'Content-type: text/xml; charset="$Charset"',
  '_start' => '<?xml version="1.0" encoding="$Charset"?'.'>
<rdf:RDF xmlns="http://purl.org/rss/1.0/"
         xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
         xmlns:dc="http://purl.org/dc/elements/1.1/">
  <channel rdf:about="{$PageUrl}?action=rdf">'."\n",
  'title' => '$WikiTitle | {$Group} / {$Title}',
  'link' => '{$PageUrl}?action=rdf',
  'description' => '{$Group}.{$Title}',
  'dc:date' => '$FeedISOTime',
  'items' => "<items>\n<rdf:Seq>\n\$FeedRDFSeq</rdf:Seq>\n</items>\n",
  '_items' => "</channel>\n",
  '_end' => "</rdf:RDF>\n"));
SDVA($FeedFmt['rdf']['item'], array(
  '_start' => "<item rdf:about=\"{\$PageUrl}\">\n",
  '_end' => "</item>\n",
  'title' => '$WikiTitle | {$Group} / {$Title}',
  'link' => '{$PageUrl}',
  'description' => '{$Description}',
  'dc:date' => '$ItemISOTime'));
  
foreach(array_keys($FeedFmt) as $k) {
  SDV($HandleActions[$k], 'HandleFeed');
  SDV($HandleAuth[$k], 'read');
}

function HandleFeed($pagename, $auth = 'read') {
  global $FeedFmt, $action, $PCache, $FmtV, $TimeISOZFmt, $RSSTimeFmt,
    $FeedPageListOpt, $FeedCategoryOpt, $FeedTrailOpt,
    $FeedDescPatterns, $CategoryGroup, $EntitiesTable;
  SDV($RSSTimeFmt, 'D, d M Y H:i:s \G\M\T');
  SDV($FeedDescPatterns, 
    array('/<[^>]*$/' => ' ', '/\\w+$/' => '', '/<[^>]+>/' => ''));
  $FeedPageListOpt = (array)@$FeedPageListOpt;
  SDVA($FeedCategoryOpt, array('link' => $pagename));
  SDVA($FeedTrailOpt, array('trail' => $pagename, 'count' => 10));

  $f = $FeedFmt[$action];
  $page = RetrieveAuthPage($pagename, $auth, true, READPAGE_CURRENT);
  if (!$page) Abort("?cannot generate feed");
  $feedtime = $page['time'];

  # determine list of pages to display
  if (@($_REQUEST['trail'] || $_REQUEST['group'] || $_REQUEST['link'] 
        || $_REQUEST['name'])) 
    $opt = $FeedPageListOpt;
  else if (preg_match("/^$CategoryGroup\\./", $pagename)) 
    $opt = $FeedCategoryOpt;
  else if ($action != 'dc') $opt = $FeedTrailOpt;
  else { 
    PCache($pagename, $page); 
    $pagelist = array($pagename); 
  }
  if (!$pagelist) {
    $opt = array_merge($opt, @$_REQUEST);
    $pagelist = MakePageList($pagename, $opt, 0);
  }

  # process list of pages in feed
  $rdfseq = '';
  $pl = array();
  foreach($pagelist as $pn) {
    if (!PageExists($pn)) continue;
    if (!isset($PCache[$pn]['time'])) 
      { $page = ReadPage($pn, READPAGE_CURRENT); PCache($pn, $page); }
    $pc = & $PCache[$pn];
    $pl[] = $pn;
    if (@$opt['count'] && count($pl) >= $opt['count']) break;
    $rdfseq .= FmtPageName("<rdf:li resource=\"{\$PageUrl}\" />\n", $pn);
    if ($pc['time'] > $feedtime) $feedtime = $pc['time'];
  }
  $pagelist = $pl;

  $FmtV['$FeedRDFSeq'] = $rdfseq;
  $FmtV['$FeedISOTime'] = gmstrftime($TimeISOZFmt, $feedtime);
  $FmtV['$FeedRSSTime'] = gmdate($RSSTimeFmt, $feedtime);
  # format start of feed
  $out = FmtPageName($f['feed']['_start'], $pagename);

  # format feed elements
  foreach($f['feed'] as $k => $v) {
    if ($k{0} == '_' || !$v) continue;
    $x = FmtPageName($v, $pagename);
    if (!$x) continue;
    $out .= ($v{0} == '<') ? $x : "<$k>$x</$k>\n";
  }

  # format items in feed
  if (@$f['feed']['_items']) 
    $out .= FmtPageName($f['feed']['_items'], $pagename);
  foreach($pagelist as $pn) {
    $page = &$PCache[$pn];
    $FmtV['$ItemDesc'] = @$page['description'];
    $FmtV['$ItemISOTime'] = gmstrftime($TimeISOZFmt, $page['time']);
    $FmtV['$ItemRSSTime'] = gmdate($RSSTimeFmt, $page['time']);

    $out .= FmtPageName($f['item']['_start'], $pn);
    foreach((array)@$f['item'] as $k => $v) {
      if ($k{0} == '_' || !$v) continue;
      if (is_callable($v)) { $out .= $v($pn, $page, $k); continue; }
      if (strpos($v, '$LastModifiedBy') !== false && !@$page['author']) 
        continue;
      if (strpos($v, '$Category') !== false) {
        if (preg_match_all("/(?<=^|,)$CategoryGroup\\.([^,]+)/", 
                           @$page['targets'], $match)) {
          foreach($match[1] as $c) {
            $FmtV['$Category'] = $c;
            $out .= FmtPageName($v, $pn);
          }
        }
        continue;
      }
      $x = FmtPageName($v, $pn);
      if (!$x) continue;
      $out .= ($v{0} == '<') ? $x : "<$k>$x</$k>\n";
    }
    $out .= FmtPageName($f['item']['_end'], $pn);
  } 
  $out .= FmtPageName($f['feed']['_end'], $pagename);
  foreach((array)@$f['feed']['_header'] as $fmt)
    header(FmtPageName($fmt, $pagename));
  print str_replace(array_keys($EntitiesTable),
                    array_values($EntitiesTable), $out);
}

## RSSEnclosure is called in ?action=rss to generate <enclosure>
## tags for any pages that have an attached "PageName.mp3" file.
## The set of attachments to enclose is given by $RSSEnclosureFmt.
function RSSEnclosure($pagename, &$page, $k) {
  global $RSSEnclosureFmt, $UploadFileFmt, $UploadExts;
  if (!function_exists('MakeUploadName')) return '';
  SDV($RSSEnclosureFmt, array('{$Name}.mp3'));
  $encl = '';
  foreach((array)$RSSEnclosureFmt as $fmt) {
    $path = FmtPageName($fmt, $pagename);
    $upname = MakeUploadName($pagename, $path);
    $filepath = FmtPageName("$UploadFileFmt/$upname", $pagename);
    if (file_exists($filepath)) {
      $length = filesize($filepath);
      $type = @$UploadExts[preg_replace('/.*\\./', '', $filepath)];
      $url = LinkUpload($pagename, 'Attach:', $path, '', '', '$LinkUrl');
      $encl .= "<$k url='$url' length='$length' type='$type' />";
    }
  }
  return $encl;
}

## Since most feeds don't understand html character entities, we
## convert the common ones to their numeric form here.
SDVA($EntitiesTable, array(
  # entities defined in "http://www.w3.org/TR/xhtml1/DTD/xhtml-lat1.ent"
  '&nbsp;' => '&#160;', 
  '&iexcl;' => '&#161;', 
  '&cent;' => '&#162;', 
  '&pound;' => '&#163;', 
  '&curren;' => '&#164;', 
  '&yen;' => '&#165;', 
  '&brvbar;' => '&#166;', 
  '&sect;' => '&#167;', 
  '&uml;' => '&#168;', 
  '&copy;' => '&#169;', 
  '&ordf;' => '&#170;', 
  '&laquo;' => '&#171;', 
  '&not;' => '&#172;', 
  '&shy;' => '&#173;', 
  '&reg;' => '&#174;', 
  '&macr;' => '&#175;', 
  '&deg;' => '&#176;', 
  '&plusmn;' => '&#177;', 
  '&sup2;' => '&#178;', 
  '&sup3;' => '&#179;', 
  '&acute;' => '&#180;', 
  '&micro;' => '&#181;', 
  '&para;' => '&#182;', 
  '&middot;' => '&#183;', 
  '&cedil;' => '&#184;', 
  '&sup1;' => '&#185;', 
  '&ordm;' => '&#186;', 
  '&raquo;' => '&#187;', 
  '&frac14;' => '&#188;', 
  '&frac12;' => '&#189;', 
  '&frac34;' => '&#190;', 
  '&iquest;' => '&#191;', 
  '&Agrave;' => '&#192;', 
  '&Aacute;' => '&#193;', 
  '&Acirc;' => '&#194;', 
  '&Atilde;' => '&#195;', 
  '&Auml;' => '&#196;', 
  '&Aring;' => '&#197;', 
  '&AElig;' => '&#198;', 
  '&Ccedil;' => '&#199;', 
  '&Egrave;' => '&#200;', 
  '&Eacute;' => '&#201;', 
  '&Ecirc;' => '&#202;', 
  '&Euml;' => '&#203;', 
  '&Igrave;' => '&#204;', 
  '&Iacute;' => '&#205;', 
  '&Icirc;' => '&#206;', 
  '&Iuml;' => '&#207;', 
  '&ETH;' => '&#208;', 
  '&Ntilde;' => '&#209;', 
  '&Ograve;' => '&#210;', 
  '&Oacute;' => '&#211;', 
  '&Ocirc;' => '&#212;', 
  '&Otilde;' => '&#213;', 
  '&Ouml;' => '&#214;', 
  '&times;' => '&#215;', 
  '&Oslash;' => '&#216;', 
  '&Ugrave;' => '&#217;', 
  '&Uacute;' => '&#218;', 
  '&Ucirc;' => '&#219;', 
  '&Uuml;' => '&#220;', 
  '&Yacute;' => '&#221;', 
  '&THORN;' => '&#222;', 
  '&szlig;' => '&#223;', 
  '&agrave;' => '&#224;', 
  '&aacute;' => '&#225;', 
  '&acirc;' => '&#226;', 
  '&atilde;' => '&#227;', 
  '&auml;' => '&#228;', 
  '&aring;' => '&#229;', 
  '&aelig;' => '&#230;', 
  '&ccedil;' => '&#231;', 
  '&egrave;' => '&#232;', 
  '&eacute;' => '&#233;', 
  '&ecirc;' => '&#234;', 
  '&euml;' => '&#235;', 
  '&igrave;' => '&#236;', 
  '&iacute;' => '&#237;', 
  '&icirc;' => '&#238;', 
  '&iuml;' => '&#239;', 
  '&eth;' => '&#240;', 
  '&ntilde;' => '&#241;', 
  '&ograve;' => '&#242;', 
  '&oacute;' => '&#243;', 
  '&ocirc;' => '&#244;', 
  '&otilde;' => '&#245;', 
  '&ouml;' => '&#246;', 
  '&divide;' => '&#247;', 
  '&oslash;' => '&#248;', 
  '&ugrave;' => '&#249;', 
  '&uacute;' => '&#250;', 
  '&ucirc;' => '&#251;', 
  '&uuml;' => '&#252;', 
  '&yacute;' => '&#253;', 
  '&thorn;' => '&#254;', 
  '&yuml;' => '&#255;', 
  # entities defined in "http://www.w3.org/TR/xhtml1/DTD/xhtml-special.ent"
  '&quot;' => '&#34;', 
  #'&amp;' => '&#38;#38;', 
  #'&lt;' => '&#38;#60;', 
  #'&gt;' => '&#62;', 
  '&apos;' => '&#39;', 
  '&OElig;' => '&#338;', 
  '&oelig;' => '&#339;', 
  '&Scaron;' => '&#352;', 
  '&scaron;' => '&#353;', 
  '&Yuml;' => '&#376;', 
  '&circ;' => '&#710;', 
  '&tilde;' => '&#732;', 
  '&ensp;' => '&#8194;', 
  '&emsp;' => '&#8195;', 
  '&thinsp;' => '&#8201;', 
  '&zwnj;' => '&#8204;', 
  '&zwj;' => '&#8205;', 
  '&lrm;' => '&#8206;', 
  '&rlm;' => '&#8207;', 
  '&ndash;' => '&#8211;', 
  '&mdash;' => '&#8212;', 
  '&lsquo;' => '&#8216;', 
  '&rsquo;' => '&#8217;', 
  '&sbquo;' => '&#8218;', 
  '&ldquo;' => '&#8220;', 
  '&rdquo;' => '&#8221;', 
  '&bdquo;' => '&#8222;', 
  '&dagger;' => '&#8224;', 
  '&Dagger;' => '&#8225;', 
  '&permil;' => '&#8240;', 
  '&lsaquo;' => '&#8249;', 
  '&rsaquo;' => '&#8250;', 
  '&euro;' => '&#8364;', 
  # entities defined in "http://www.w3.org/TR/xhtml1/DTD/xhtml-symbol.ent"
  '&fnof;' => '&#402;', 
  '&Alpha;' => '&#913;', 
  '&Beta;' => '&#914;', 
  '&Gamma;' => '&#915;', 
  '&Delta;' => '&#916;', 
  '&Epsilon;' => '&#917;', 
  '&Zeta;' => '&#918;', 
  '&Eta;' => '&#919;', 
  '&Theta;' => '&#920;', 
  '&Iota;' => '&#921;', 
  '&Kappa;' => '&#922;', 
  '&Lambda;' => '&#923;', 
  '&Mu;' => '&#924;', 
  '&Nu;' => '&#925;', 
  '&Xi;' => '&#926;', 
  '&Omicron;' => '&#927;', 
  '&Pi;' => '&#928;', 
  '&Rho;' => '&#929;', 
  '&Sigma;' => '&#931;', 
  '&Tau;' => '&#932;', 
  '&Upsilon;' => '&#933;', 
  '&Phi;' => '&#934;', 
  '&Chi;' => '&#935;', 
  '&Psi;' => '&#936;', 
  '&Omega;' => '&#937;', 
  '&alpha;' => '&#945;', 
  '&beta;' => '&#946;', 
  '&gamma;' => '&#947;', 
  '&delta;' => '&#948;', 
  '&epsilon;' => '&#949;', 
  '&zeta;' => '&#950;', 
  '&eta;' => '&#951;', 
  '&theta;' => '&#952;', 
  '&iota;' => '&#953;', 
  '&kappa;' => '&#954;', 
  '&lambda;' => '&#955;', 
  '&mu;' => '&#956;', 
  '&nu;' => '&#957;', 
  '&xi;' => '&#958;', 
  '&omicron;' => '&#959;', 
  '&pi;' => '&#960;', 
  '&rho;' => '&#961;', 
  '&sigmaf;' => '&#962;', 
  '&sigma;' => '&#963;', 
  '&tau;' => '&#964;', 
  '&upsilon;' => '&#965;', 
  '&phi;' => '&#966;', 
  '&chi;' => '&#967;', 
  '&psi;' => '&#968;', 
  '&omega;' => '&#969;', 
  '&thetasym;' => '&#977;', 
  '&upsih;' => '&#978;', 
  '&piv;' => '&#982;', 
  '&bull;' => '&#8226;', 
  '&hellip;' => '&#8230;', 
  '&prime;' => '&#8242;', 
  '&Prime;' => '&#8243;', 
  '&oline;' => '&#8254;', 
  '&frasl;' => '&#8260;', 
  '&weierp;' => '&#8472;', 
  '&image;' => '&#8465;', 
  '&real;' => '&#8476;', 
  '&trade;' => '&#8482;', 
  '&alefsym;' => '&#8501;', 
  '&larr;' => '&#8592;', 
  '&uarr;' => '&#8593;', 
  '&rarr;' => '&#8594;', 
  '&darr;' => '&#8595;', 
  '&harr;' => '&#8596;', 
  '&crarr;' => '&#8629;', 
  '&lArr;' => '&#8656;', 
  '&uArr;' => '&#8657;', 
  '&rArr;' => '&#8658;', 
  '&dArr;' => '&#8659;', 
  '&hArr;' => '&#8660;', 
  '&forall;' => '&#8704;', 
  '&part;' => '&#8706;', 
  '&exist;' => '&#8707;', 
  '&empty;' => '&#8709;', 
  '&nabla;' => '&#8711;', 
  '&isin;' => '&#8712;', 
  '&notin;' => '&#8713;', 
  '&ni;' => '&#8715;', 
  '&prod;' => '&#8719;', 
  '&sum;' => '&#8721;', 
  '&minus;' => '&#8722;', 
  '&lowast;' => '&#8727;', 
  '&radic;' => '&#8730;', 
  '&prop;' => '&#8733;', 
  '&infin;' => '&#8734;', 
  '&ang;' => '&#8736;', 
  '&and;' => '&#8743;', 
  '&or;' => '&#8744;', 
  '&cap;' => '&#8745;', 
  '&cup;' => '&#8746;', 
  '&int;' => '&#8747;', 
  '&there4;' => '&#8756;', 
  '&sim;' => '&#8764;', 
  '&cong;' => '&#8773;', 
  '&asymp;' => '&#8776;', 
  '&ne;' => '&#8800;', 
  '&equiv;' => '&#8801;', 
  '&le;' => '&#8804;', 
  '&ge;' => '&#8805;', 
  '&sub;' => '&#8834;', 
  '&sup;' => '&#8835;', 
  '&nsub;' => '&#8836;', 
  '&sube;' => '&#8838;', 
  '&supe;' => '&#8839;', 
  '&oplus;' => '&#8853;', 
  '&otimes;' => '&#8855;', 
  '&perp;' => '&#8869;', 
  '&sdot;' => '&#8901;', 
  '&lceil;' => '&#8968;', 
  '&rceil;' => '&#8969;', 
  '&lfloor;' => '&#8970;', 
  '&rfloor;' => '&#8971;', 
  '&lang;' => '&#9001;', 
  '&rang;' => '&#9002;', 
  '&loz;' => '&#9674;', 
  '&spades;' => '&#9824;', 
  '&clubs;' => '&#9827;', 
  '&hearts;' => '&#9829;', 
  '&diams;' => '&#9830;'));

